File: internal/protocol/codes.go
package protocol

// =============================================================================
// Error Codes — machine-readable identifiers for all protocol errors.
// These are sent over the wire (JSON). NEVER change existing codes — only add new ones.
// Format: CATEGORY_SPECIFIC_DETAIL (SCREAMING_SNAKE_CASE)
// =============================================================================

// --- General ---
const (
	CodeOK                = "OK"
	CodeUnknownError      = "UNKNOWN_ERROR"
	CodeInvalidJSON       = "INVALID_JSON"
	CodeInvalidPayload    = "INVALID_PAYLOAD"
	CodeAgentNotFound     = "AGENT_NOT_FOUND"
	CodeAgentNotConnected = "AGENT_NOT_CONNECTED"
	CodeUnauthorized      = "UNAUTHORIZED"
	CodeForbidden         = "FORBIDDEN"
	CodeInternalError     = "INTERNAL_ERROR"
	CodeUnknownMessage    = "UNKNOWN_MESSAGE_TYPE"
)

// --- Authentication ---
const (
	CodeTokenMissing       = "TOKEN_MISSING"
	CodeTokenInvalid       = "TOKEN_INVALID"
	CodeTokenExpired       = "TOKEN_EXPIRED"
	CodeTokenRevoked       = "TOKEN_REVOKED"
	CodeTokenBlacklisted   = "TOKEN_BLACKLISTED"
	CodeSignatureInvalid   = "SIGNATURE_INVALID"
	CodeSecretNotFound     = "SECRET_NOT_FOUND"
	CodeTokenTypeInvalid   = "TOKEN_TYPE_INVALID"
	CodeTokenDecodeFailed  = "TOKEN_DECODE_FAILED"
	CodeTokenParseFailed   = "TOKEN_PARSE_FAILED"
	CodeTokenSerializeFail = "TOKEN_SERIALIZE_FAILED"
)

// --- Agent Lifecycle ---
const (
	CodeAgentConnectFailed = "AGENT_CONNECT_FAILED"
	CodeAgentDisconnected  = "AGENT_DISCONNECTED"
	CodeAgentReadFailed    = "AGENT_READ_FAILED"
	CodeAgentWriteFailed   = "AGENT_WRITE_FAILED"
	CodeAgentNotAuthorized = "AGENT_NOT_AUTHORIZED"
	CodeAgentLimitExceeded = "AGENT_LIMIT_EXCEEDED"
	CodeAgentPaused        = "AGENT_PAUSED"
	CodeAgentEmergencyStop = "AGENT_EMERGENCY_STOP"
)

// --- Execution ---
const (
	CodeExecSuccess         = "EXEC_SUCCESS"
	CodeExecTimeout         = "EXEC_TIMEOUT"
	CodeExecBlocked         = "EXEC_BLOCKED"
	CodeExecBlockedReadOnly = "EXEC_BLOCKED_READONLY"
	CodeExecBlockedSandbox  = "EXEC_BLOCKED_SANDBOX"
	CodeExecBlockedSudo     = "EXEC_BLOCKED_SUDO"
	CodeExecFailed          = "EXEC_FAILED"
	CodeExecNeedsApproval   = "EXEC_NEEDS_APPROVAL"
	CodeExecAwaitingApproval = "EXEC_AWAITING_APPROVAL"
	CodeExecRejected        = "EXEC_REJECTED"
	CodeExecApproved        = "EXEC_APPROVED"
)

// --- File Operations ---
const (
	CodeFileEmptyPath     = "FILE_EMPTY_PATH"
	CodeFileInvalidPath   = "FILE_INVALID_PATH"
	CodeFileNotFound      = "FILE_NOT_FOUND"
	CodeFileTooLarge      = "FILE_TOO_LARGE"
	CodeFileReadError     = "FILE_READ_ERROR"
	CodeFileWriteError    = "FILE_WRITE_ERROR"
	CodeFileDecodeError   = "FILE_DECODE_ERROR"
	CodeFileDirCreateError = "FILE_DIR_CREATE_ERROR"
	CodeFileParentDirError = "FILE_PARENT_DIR_ERROR"
	CodeFileDirReadError  = "FILE_DIR_READ_ERROR"
)

// --- Config ---
const (
	CodeConfigApplied   = "CONFIG_APPLIED"
	CodeConfigFailed    = "CONFIG_UPDATE_FAILED"
	CodeConfigLoadError = "CONFIG_LOAD_ERROR"
	CodeConfigSaveError = "CONFIG_SAVE_ERROR"
	CodeConfigParseError = "CONFIG_PARSE_ERROR"
)

// --- Skills ---
const (
	CodeSkillAlreadyExists = "SKILL_ALREADY_EXISTS"
	CodeSkillNotFound      = "SKILL_NOT_FOUND"
	CodeSkillSaveError     = "SKILL_SAVE_ERROR"
	CodeSkillDeleteError   = "SKILL_DELETE_ERROR"
	CodeSkillDirError      = "SKILL_DIR_CREATE_ERROR"
	CodeSkillSerializeError = "SKILL_SERIALIZE_ERROR"
)

// --- Tasks (Autonomous) ---
const (
	CodeTaskAccepted  = "TASK_ACCEPTED"
	CodeTaskError     = "TASK_ERROR"
	CodeTaskCancelError = "TASK_CANCEL_ERROR"
	CodeTaskStepStart = "TASK_STEP_START"
	CodeTaskStepDone  = "TASK_STEP_DONE"
	CodeTaskDone      = "TASK_DONE"
	CodeTaskErrorDone = "TASK_ERROR_DONE"
)

// --- Kill Switch ---
const (
	CodeKillSwitchDiskFull = "KILL_SWITCH_DISK_FULL"
	CodeKillSwitchCPUHigh  = "KILL_SWITCH_CPU_HIGH"
	CodeKillSwitchPause    = "KILL_SWITCH_PAUSED"
	CodeKillSwitchResume   = "KILL_SWITCH_RESUMED"
	CodeKillSwitchEmergency = "KILL_SWITCH_EMERGENCY"
)

// --- Backup ---
const (
	CodeBackupEmptyPaths     = "BACKUP_EMPTY_PATHS"
	CodeBackupCreateError    = "BACKUP_CREATE_ERROR"
	CodeBackupArchiveError   = "BACKUP_ARCHIVE_ERROR"
	CodeBackupSnapshotNotFound = "BACKUP_SNAPSHOT_NOT_FOUND"
	CodeBackupRestoreError   = "BACKUP_RESTORE_ERROR"
	CodeBackupRestoreOpenError = "BACKUP_RESTORE_OPEN_ERROR"
	CodeBackupRestoreGzipError = "BACKUP_RESTORE_GZIP_ERROR"
	CodeBackupRestoreReadError = "BACKUP_RESTORE_READ_ERROR"
	CodeBackupRestoreDirError  = "BACKUP_RESTORE_DIR_ERROR"
	CodeBackupRestoreFileError = "BACKUP_RESTORE_FILE_ERROR"
	CodeBackupDeleteError   = "BACKUP_DELETE_ERROR"
	CodeBackupMetadataError  = "BACKUP_METADATA_ERROR"
	CodeBackupDirCreateError = "BACKUP_DIR_CREATE_ERROR"
	CodeBackupSerializeError = "BACKUP_SERIALIZE_ERROR"
	CodeBackupGlobError     = "BACKUP_GLOB_ERROR"
	CodeBackupFileAddError  = "BACKUP_FILE_ADD_ERROR"
	CodeBackupCleanup       = "BACKUP_CLEANUP"
)

// --- Audit ---
const (
	CodeAuditDirCreateError = "AUDIT_DIR_CREATE_ERROR"
	CodeAuditFileOpenError  = "AUDIT_FILE_OPEN_ERROR"
	CodeAuditSerializeError = "AUDIT_SERIALIZE_ERROR"
	CodeAuditWriteError     = "AUDIT_WRITE_ERROR"
	CodeAuditFormatUnsupported = "AUDIT_FORMAT_UNSUPPORTED"
)

// --- Registry ---
const (
	CodeClientNotFound    = "CLIENT_NOT_FOUND"
	CodeClientDeactivated = "CLIENT_DEACTIVATED"
	CodeClientCreateError = "CLIENT_CREATE_ERROR"
	CodeClientSaveError   = "CLIENT_SAVE_ERROR"
	CodeClientLoadError   = "CLIENT_LOAD_ERROR"
	CodeRegistryLoadError = "REGISTRY_LOAD_ERROR"
	CodeRegistrySaveError = "REGISTRY_SAVE_ERROR"
	CodeTokenGenerateError = "TOKEN_GENERATE_ERROR"
)

// --- TLS ---
const (
	CodeTLSKeyGenerateError    = "TLS_KEY_GENERATE_ERROR"
	CodeTLSSerialGenerateError = "TLS_SERIAL_GENERATE_ERROR"
	CodeTLSCertCreateError     = "TLS_CERT_CREATE_ERROR"
	CodeTLSCertDirError        = "TLS_CERT_DIR_CREATE_ERROR"
	CodeTLSCertWriteError      = "TLS_CERT_WRITE_ERROR"
	CodeTLSCertLoadError       = "TLS_CERT_LOAD_ERROR"
	CodeTLSCertMissing         = "TLS_CERT_MISSING"
	CodeTLSModeUnknown         = "TLS_MODE_UNKNOWN"
)

// --- MCP ---
const (
	CodeMCPAgentNotFound  = "MCP_AGENT_NOT_FOUND"
	CodeMCPTimeout        = "MCP_TIMEOUT"
	CodeMCPAgentError     = "MCP_AGENT_ERROR"
)

// --- LLM Proxy ---
const (
	CodeLLMAllBackendsDown = "LLM_ALL_BACKENDS_DOWN"
	CodeLLMRequestError    = "LLM_REQUEST_ERROR"
	CodeLLMResponseError   = "LLM_RESPONSE_ERROR"
	CodeLLMEmptyResponse   = "LLM_EMPTY_RESPONSE"
	CodeLLMParseError      = "LLM_PARSE_ERROR"
)

File: crates/core/src/codes.rs
// FlowLink Core — Error codes
// Port of internal/protocol/codes.go
// NEVER change existing codes — only add new ones.

pub mod codes {
    // General
    pub const OK: &str = "OK";
    pub const UNKNOWN_ERROR: &str = "UNKNOWN_ERROR";
    pub const INVALID_JSON: &str = "INVALID_JSON";
    pub const INVALID_PAYLOAD: &str = "INVALID_PAYLOAD";
    pub const AGENT_NOT_FOUND: &str = "AGENT_NOT_FOUND";
    pub const AGENT_NOT_CONNECTED: &str = "AGENT_NOT_CONNECTED";
    pub const UNAUTHORIZED: &str = "UNAUTHORIZED";
    pub const FORBIDDEN: &str = "FORBIDDEN";
    pub const INTERNAL_ERROR: &str = "INTERNAL_ERROR";
    pub const UNKNOWN_MESSAGE_TYPE: &str = "UNKNOWN_MESSAGE_TYPE";

    // Authentication
    pub const TOKEN_MISSING: &str = "TOKEN_MISSING";
    pub const TOKEN_INVALID: &str = "TOKEN_INVALID";
    pub const TOKEN_EXPIRED: &str = "TOKEN_EXPIRED";
    pub const TOKEN_REVOKED: &str = "TOKEN_REVOKED";
    pub const TOKEN_BLACKLISTED: &str = "TOKEN_BLACKLISTED";
    pub const SIGNATURE_INVALID: &str = "SIGNATURE_INVALID";
    pub const SECRET_NOT_FOUND: &str = "SECRET_NOT_FOUND";

    // Agent Lifecycle
    pub const AGENT_CONNECT_FAILED: &str = "AGENT_CONNECT_FAILED";
    pub const AGENT_DISCONNECTED: &str = "AGENT_DISCONNECTED";
    pub const AGENT_NOT_AUTHORIZED: &str = "AGENT_NOT_AUTHORIZED";
    pub const AGENT_LIMIT_EXCEEDED: &str = "AGENT_LIMIT_EXCEEDED";
    pub const AGENT_PAUSED: &str = "AGENT_PAUSED";
    pub const AGENT_EMERGENCY_STOP: &str = "AGENT_EMERGENCY_STOP";
    pub const PROTOCOL_VERSION_MISMATCH: &str = "PROTOCOL_VERSION_MISMATCH";

    // Execution
    pub const EXEC_SUCCESS: &str = "EXEC_SUCCESS";
    pub const EXEC_TIMEOUT: &str = "EXEC_TIMEOUT";
    pub const EXEC_BLOCKED: &str = "EXEC_BLOCKED";
    pub const EXEC_BLOCKED_READONLY: &str = "EXEC_BLOCKED_READONLY";
    pub const EXEC_BLOCKED_SANDBOX: &str = "EXEC_BLOCKED_SANDBOX";
    pub const EXEC_BLOCKED_SUDO: &str = "EXEC_BLOCKED_SUDO";
    pub const EXEC_FAILED: &str = "EXEC_FAILED";
    pub const EXEC_NEEDS_APPROVAL: &str = "EXEC_NEEDS_APPROVAL";
    pub const EXEC_AWAITING_APPROVAL: &str = "EXEC_AWAITING_APPROVAL";
    pub const EXEC_REJECTED: &str = "EXEC_REJECTED";
    pub const EXEC_APPROVED: &str = "EXEC_APPROVED";

    // File Operations
    pub const FILE_EMPTY_PATH: &str = "FILE_EMPTY_PATH";
    pub const FILE_INVALID_PATH: &str = "FILE_INVALID_PATH";
    pub const FILE_NOT_FOUND: &str = "FILE_NOT_FOUND";
    pub const FILE_TOO_LARGE: &str = "FILE_TOO_LARGE";
    pub const FILE_READ_ERROR: &str = "FILE_READ_ERROR";
    pub const FILE_WRITE_ERROR: &str = "FILE_WRITE_ERROR";

    // Config
    pub const CONFIG_APPLIED: &str = "CONFIG_APPLIED";
    pub const CONFIG_FAILED: &str = "CONFIG_UPDATE_FAILED";
    pub const CONFIG_LOAD_ERROR: &str = "CONFIG_LOAD_ERROR";

    // Skills
    pub const SKILL_ALREADY_EXISTS: &str = "SKILL_ALREADY_EXISTS";
    pub const SKILL_NOT_FOUND: &str = "SKILL_NOT_FOUND";

    // Tasks (Autonomous)
    pub const TASK_ACCEPTED: &str = "TASK_ACCEPTED";
    pub const TASK_ERROR: &str = "TASK_ERROR";
    pub const TASK_DONE: &str = "TASK_DONE";

    // Kill Switch
    pub const KILL_SWITCH_DISK_FULL: &str = "KILL_SWITCH_DISK_FULL";
    pub const KILL_SWITCH_CPU_HIGH: &str = "KILL_SWITCH_CPU_HIGH";
    pub const KILL_SWITCH_PAUSED: &str = "KILL_SWITCH_PAUSED";
    pub const KILL_SWITCH_RESUMED: &str = "KILL_SWITCH_RESUMED";
    pub const KILL_SWITCH_EMERGENCY: &str = "KILL_SWITCH_EMERGENCY";

    // Backup
    pub const BACKUP_CREATE_ERROR: &str = "BACKUP_CREATE_ERROR";
    pub const BACKUP_SNAPSHOT_NOT_FOUND: &str = "BACKUP_SNAPSHOT_NOT_FOUND";
    pub const BACKUP_RESTORE_ERROR: &str = "BACKUP_RESTORE_ERROR";
    pub const BACKUP_DELETE_ERROR: &str = "BACKUP_DELETE_ERROR";
    pub const BACKUP_CHECKSUM_MISMATCH: &str = "BACKUP_CHECKSUM_MISMATCH";

    // MCP
    pub const MCP_AGENT_NOT_FOUND: &str = "MCP_AGENT_NOT_FOUND";
    pub const MCP_TIMEOUT: &str = "MCP_TIMEOUT";

    // LLM Proxy
    pub const LLM_ALL_BACKENDS_DOWN: &str = "LLM_ALL_BACKENDS_DOWN";
    pub const LLM_REQUEST_ERROR: &str = "LLM_REQUEST_ERROR";

    // Shield (NEW v2)
    pub const SHIELD_BLOCKED: &str = "SHIELD_BLOCKED";
    pub const SHIELD_SNAPSHOT_CREATED: &str = "SHIELD_SNAPSHOT_CREATED";
    pub const SHIELD_PROCESS_STOPPED: &str = "SHIELD_PROCESS_STOPPED";
    pub const SHIELD_PROCESS_KILLED: &str = "SHIELD_PROCESS_KILLED";
    pub const SHIELD_TIMEOUT: &str = "SHIELD_TIMEOUT";
    pub const SHIELD_ALERT_SENT: &str = "SHIELD_ALERT_SENT";
}

#[cfg(test)]
mod tests {
    use super::*;

    fn all_codes() -> Vec<&'static str> {
        vec![
            OK, UNKNOWN_ERROR, INVALID_JSON, INVALID_PAYLOAD, AGENT_NOT_FOUND,
            AGENT_NOT_CONNECTED, UNAUTHORIZED, FORBIDDEN, INTERNAL_ERROR, UNKNOWN_MESSAGE_TYPE,
            TOKEN_MISSING, TOKEN_INVALID, TOKEN_EXPIRED, TOKEN_REVOKED, TOKEN_BLACKLISTED,
            SIGNATURE_INVALID, SECRET_NOT_FOUND,
            AGENT_CONNECT_FAILED, AGENT_DISCONNECTED, AGENT_NOT_AUTHORIZED,
            AGENT_LIMIT_EXCEEDED, AGENT_PAUSED, AGENT_EMERGENCY_STOP, PROTOCOL_VERSION_MISMATCH,
            EXEC_SUCCESS, EXEC_TIMEOUT, EXEC_BLOCKED, EXEC_BLOCKED_READONLY, EXEC_BLOCKED_SANDBOX,
            EXEC_BLOCKED_SUDO, EXEC_FAILED, EXEC_NEEDS_APPROVAL, EXEC_AWAITING_APPROVAL,
            EXEC_REJECTED, EXEC_APPROVED,
            FILE_EMPTY_PATH, FILE_INVALID_PATH, FILE_NOT_FOUND, FILE_TOO_LARGE,
            FILE_READ_ERROR, FILE_WRITE_ERROR,
            CONFIG_APPLIED, CONFIG_FAILED, CONFIG_LOAD_ERROR,
            SKILL_ALREADY_EXISTS, SKILL_NOT_FOUND,
            TASK_ACCEPTED, TASK_ERROR, TASK_DONE,
            KILL_SWITCH_DISK_FULL, KILL_SWITCH_CPU_HIGH, KILL_SWITCH_PAUSED,
            KILL_SWITCH_RESUMED, KILL_SWITCH_EMERGENCY,
            BACKUP_CREATE_ERROR, BACKUP_SNAPSHOT_NOT_FOUND, BACKUP_RESTORE_ERROR,
            BACKUP_DELETE_ERROR, BACKUP_CHECKSUM_MISMATCH,
            MCP_AGENT_NOT_FOUND, MCP_TIMEOUT,
            LLM_ALL_BACKENDS_DOWN, LLM_REQUEST_ERROR,
            SHIELD_BLOCKED, SHIELD_SNAPSHOT_CREATED, SHIELD_PROCESS_STOPPED,
            SHIELD_PROCESS_KILLED, SHIELD_TIMEOUT, SHIELD_ALERT_SENT,
        ]
    }

    #[test]
    fn test_all_codes_are_non_empty() {
        for code in all_codes() {
            assert!(!code.is_empty(), "Code should not be empty");
        }
    }

    #[test]
    fn test_no_duplicate_codes() {
        let codes = all_codes();
        let mut seen = std::collections::HashSet::new();
        for code in &codes {
            assert!(seen.insert(*code), "Duplicate code: {}", code);
        }
    }

    #[test]
    fn test_auth_codes_exist() {
        assert_eq!(TOKEN_MISSING, "TOKEN_MISSING");
        assert_eq!(TOKEN_INVALID, "TOKEN_INVALID");
        assert_eq!(TOKEN_EXPIRED, "TOKEN_EXPIRED");
        assert_eq!(TOKEN_REVOKED, "TOKEN_REVOKED");
    }

    #[test]
    fn test_exec_codes_exist() {
        assert_eq!(EXEC_SUCCESS, "EXEC_SUCCESS");
        assert_eq!(EXEC_TIMEOUT, "EXEC_TIMEOUT");
        assert_eq!(EXEC_BLOCKED, "EXEC_BLOCKED");
        assert_eq!(EXEC_FAILED, "EXEC_FAILED");
    }

    #[test]
    fn test_file_codes_exist() {
        assert_eq!(FILE_NOT_FOUND, "FILE_NOT_FOUND");
        assert_eq!(FILE_TOO_LARGE, "FILE_TOO_LARGE");
        assert_eq!(FILE_EMPTY_PATH, "FILE_EMPTY_PATH");
    }

    #[test]
    fn test_backup_codes_exist() {
        assert_eq!(BACKUP_CREATE_ERROR, "BACKUP_CREATE_ERROR");
        assert_eq!(BACKUP_RESTORE_ERROR, "BACKUP_RESTORE_ERROR");
        assert_eq!(BACKUP_CHECKSUM_MISMATCH, "BACKUP_CHECKSUM_MISMATCH");
    }
}
